File: Data_Structure/Non_Linear/HashSet/C++/HashSet.cpp
#include <bits/stdc++.h>
using namespace std;
template <typename k>
class HashSet{
private:
	vector<vector<k>> vec;
	int length;
	int count;
public:
	HashSet(){
		this->length = 16;
		vec.resize(this->length);
		this->count = 0;
	}

	size_t hashCode(k key){
		hash<k> h;
		return h(key);
	}
	void put(k key){
		int index = (int)hashCode(key) % this->length;
		int l = vec[index].size();
		for(int i = 0 ; i < l ; i++){
			if(vec[index][i] == key){
				return;
			}
		}
		vec[index].emplace_back(key);
		this->count++;
	}
	k get(k key){
		int index = (int)hashCode(key) % this->length;
		for(auto a : vec[index]){
			if(a == key){
				return a;
			}
		}
		return -1;
	}
	bool remove(k key){
		int index = (int)(hashCode(key)) % this->length;
		int l = vec[index].size();
		for(int i = 0 ; i < l ;i++){
			if(vec[index][i] == key){
				int j = i;
				while(j < l-1){
					vec[index][j]=vec[index][j+1];
				}
				vec[index].pop_back();
				this->count--;
				return true;
			}
		}
		return false;
	}
	int size(){
		return this->count;
	}
	bool isEmpty(){
		return this->count == 0;
	}
	void clear(){
		vec.clear();
		vec.resize(this->length);
		this->count = 0;
	}

};
int main(){
	HashSet<int> set;
	set.put(1);
	set.put(2);
	set.put(3);
	set.put(4);
	cout<<"----------------- testing get ----------------- \n";
	cout<<"get 1 : "<<set.get(1)<<endl;
	cout<<"get 2 : "<<set.get(2)<<endl;
	cout<<"get 3 : "<<set.get(3)<<endl;
	cout<<"get 8 (not in set): "<<set.get(8)<<endl;

	cout<<"----------------- testing remove -----------------\n";
	cout<<"size : "<<set.size()<<endl;
	set.remove(1);
	cout<<"size after removing 1 : "<<set.size()<<endl;

	cout<<"----------------- testing get deleted key -----------------\n";
	cout<<"get 1 (deleted) : "<<set.get(1)<<endl;

	cout<<"----------------- testing isEmpty -----------------\n";
	cout<<"is empty : "<<(set.isEmpty() ? "true": "false")<<endl;

	cout<<"----------------- testing clear -----------------\n";
	set.clear();
	cout<<"is empty : "<<(set.isEmpty() ? "true": "false")<<endl;

	return 0;
}

File: Data_Structure/Non_Linear/HashSet/Java/HashSet.java
import java.util.*;
class HashSet <k>{
	private int count;
	private int length;
	private List<ArrayList<k>> vec;
	
	private int hashCode(k key){
		return key.hashCode();
	}
	public HashSet(){
		this.length = 16;
		vec = new ArrayList<>();
		for(int i = 0 ; i < this.length ; i++){
			vec.add(new ArrayList<k>());
		}
		this.count = 0;
	}
	public void put(k key){
		int index = (int)hashCode(key) % this.length;
		int l = vec.get(index).size();
		for(int i = 0 ; i < l ;i++){
			if(vec.get(index).get(i) == key){
				return;
			}
		}
		vec.get(index).add(key);
		this.count++;
	}
	public k get(k key){
		int index = (int)hashCode(key) % this.length;
		for(k a : vec.get(index)){
			if(hashCode(a) == hashCode(key)){
				return a;
			}
		}
		return null;
	}
	public boolean remove(k key){
		int index = (int)(hashCode(key)) % this.length;
		int l = vec.get(index).size();
		for(int i = 0 ; i < l ; i++){
			if(vec.get(index).get(i) == key){
				vec.get(index).remove(i);
				this.count--;
				return true;
			}
		}
		return false;
	}
	public boolean isEmpty() {
		return count == 0;
	}
	public int size() {
		return count;
	}
	public void clear() {
		vec = new ArrayList<>();
		for(int i = 0 ; i < this.length ; i++){
			vec.add(new ArrayList<k>());
		}
		this.count = 0;
	}
	
}